worker/raft/rafttransport/worker_test.go

```diff
@@ -159,11 +161,14 @@ func (s *WorkerSuite) SetUpTest(c *gc.C) {
 	clientTransport := s.server.Client().Transport.(*http.Transport)
 	s.config.TLSConfig = clientTransport.TLSClientConfig
 
+	s.reqsMtx.Lock()
 	s.reqs = make(chan apiserver.DetailsRequest, 10)
+	s.reqsMtx.Unlock()
+
 	s.config.Hub.Subscribe(apiserver.DetailsRequestTopic,
 		func(_ string, req apiserver.DetailsRequest, err error) {
 			c.Check(err, jc.ErrorIsNil)
-			s.reqs <- req
+			s.detailsRequest() <- req
```

Comment:
Given the look of this, I get the feeling that we aren't properly stopping the subscription in one test before the next test starts.
So TestA is getting the req #1 in the first call to SetUpTest
but then TestA "stops" and TestB 'starts" and then SetUpTest replaces the s.req1 with a new one.

But while I do think this gets rid of the race by forcing synchronization (the leftover subscription from TestA is guaranteed to see a stable reqs, but it may be the one for TestA or the one for TestB).

However, the real problem is that anything from TestA is still running by the time TestB starts, and that mean A and B are not properly isolated.


Is it possible to make sure we're fully unsubscribed by the time the TearDownTest exits?

Reply:
> Is it possible to make sure we're fully unsubscribed by the time the TearDownTest exits?

I'll look into that scenario.